Client server support

Tadinu opened the pull request:
* `ASimulationState`: Add `Server~` for service callbacks called on server only. There is a single instance owned by GameMode, which is server-only.
* `URRROS2SimulationStateClient` as UObject, moving setup of ROS2 services from `ASimulationState`. There is one created for each client's network player controller.
* Add `ARRNetworkGameMode, ARRNetworkGameState, ARRNetworkPlayerController`
* Set up `bReplicates` for `ARRBaseRobot, ASimulationState, ARRNetworkGameState, ARRNetworkPlayerController`

Commit:
working

Commit:
fixing issues with tag and target point replication

larryng-rapyuta committed:
Moved ROS2 Services to Client

larryng-rapyuta committed:
working clock sync and clock publisher/services moved to client

larryng-rapyuta committed:
working clock sync and clock publisher/services moved to client

larryng-rapyuta committed:
updated movement rpcs

larryng-rapyuta committed:
fix initial clock drift

Commit:
Merge branch 'devel' into lng_clientserver

Commit:
Refactor after merge

Commit:
Remove already lfs-added assets

Tadinu commented:
This is still wip, I just merged changes from devel and had some oss refactoring.

Commit:
RRNetworlkPlayerController::GetLifetimeReplicatedProps()
Add missing members

Commit:
RRNetworkPlayerController Rem Namespace, using ClientROS2Node->Namespace directly

Commit:
RRBaseRobot, RRRobotBaseVehicle, RobotVehicle
Turn off AutoPossessAI by default, Letting AIController auto possess in case of Non-network game mode
ARRNetworkPlayerController Add ClientInitROS2Interface()

Commit:
RRROS2GameMode add ROS2SimStateClient to init ROS2 srvs in case of NM_Standalone

yuokamoto commented:
Let us know when it become ready to review and it is better to make draft if it is not ready to be reviewed yet.

Commit:
RRROS2GameMode SimulationStateClass -> ROS2SimStateClientClass

Commit:
ARRNetworkPlayerController add CreateROS2SimStateClient(),
called in ARRNetworkGameMode::PostLogin()

Commit:
ARRNetworkPlayerController Add InitClientROS2()
WaitForPawnPossess() reorganize, ServerSetPlayerName() Add ClientROS2Node check

Commit:
ARRNetworkPlayerController
Possess any newly spawned entity that is not possessed yet

Commit:
Turn ROS2SimStateClient back to ActorComponent
So to enable its replicatability, then its creation in NetworkGameMode could be recognized in NetworkPlayerController

Commit:
Add FRREntityInfo
ASimulationState Use SpawnableEntityInfoList as replicatable

Commit:
Add more comments

Commit:
RRNetworkPlayerController
ClientROS2Node -> ServerROS2Node, ClockPublisher -> ServerClockPublisher
They are Server-only & Only their GUIDs are replicated, not Their instances

Commit:
RRNetworkPlayerController
ROS2Node->SimStateClientROS2Node, ClockPublisher->SimStateClientClockPublisher

Commit:
Add cmts on GetPawn() vs PossessedPawn

Commit:
RRNetworkPlayerController
Rollback to pawn matching in Engine

Commit:
fix build in engine

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Tools/SimulationState.h:
```diff
@@ -1,6 +1,6 @@
 /**
  * @file SimulationState.h
- * @brief SimulationState class provides ROS2 service interfaces to interact with UE4.
+ * @brief SimulationState class provides ROS2 service & action implementation to interact with UE4.
```
Do we have action?

Tadinu replied:
We don't have for now, but when we have this class should be the main interface for sim-wide ros2 communication in general,so the note is meant for that.

yuokamoto replied:
then it is better to just remove 'services' and says ROS2 interfaces, then it covers topic as well.

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Tools/SimulationState.h:
```diff
@@ -64,60 +90,53 @@ class RAPYUTASIMULATIONPLUGINS_API ASimulationState : public AActor
-     *
-     * @param InROS2Node
-     * @sa [ROS2Node](https://rclue.readthedocs.io/en/devel/doxygen_generated/html/d6/dcb/class_a_r_o_s2_node.html)
+     * @brief Fetch all entities in the current map
```
please add doxygen comments to new functions

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Robots/RobotVehicle.h:
```diff
@@ -43,12 +43,20 @@ class RAPYUTASIMULATIONPLUGINS_API ARobotVehicle : public ARRRobotBaseVehicle
     ARobotVehicle(const FObjectInitializer& ObjectInitializer);
```
out of topic of this pr
- can we make this class name SkeltalRobotVehicle?

Tadinu replied:
We should, though should be prudent since this class has BP child thus the changes would automatically invalidate them and require rework on BP classes. A separate PR should be dedicated for it.

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Core/RRNetworkGameState.cpp:
```diff
@@ -0,0 +1,28 @@
+    PrimaryActorTick.bCanEverTick = true;
+}
+
+float ARRNetworkGameState::GetServerWorldTimeSeconds() const
```
what is the intension of this function?

it returns client real time or serve rsimulation time?

larryng-rapyuta replied:
The time stored on the client's player controller (which this function retrieves) is the error corrected Server time if its on the server it just returns the server's time

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Core/RRNetworkGameState.h:
```diff
@@ -0,0 +1,23 @@
+
+public:
+    ARRNetworkGameState();
+
```
Do we need a new class just for this method?

larryng-rapyuta replied:
probably can have another way, but I saw people use gamestate for stuff like this so I followed

larryng-rapyuta replied:
also allowed for overriding the server time functions

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Tools/SimulationState.cpp:
```diff
@@ -21,212 +22,241 @@
     }
+
+    // NOTE: [SpawnableEntityInfoList] is a TArray<> thus replicatable, which is not supported for [SpawnableEntities] as a TMap
+    GetWorld()->GetTimerManager().SetTimer(TimerHandle, this, &ASimulationState::GetSpawnableEntityInfoList, 1.0f, true);
```
what this timer for?

larryng-rapyuta replied:
This one is for ensuring that the spawnable entities are replicated over, since spawnable entities doesnt have a trigger that I could borrow to use for emulating replication. The main reason for this and the EntityList 'replication' setups is due to TMaps being unable to be replicated

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Tools/SimulationState.cpp:
```diff
@@ -21,212 +22,241 @@
 }
 
-void ASimulationState::GetEntityStateSrv(UROS2GenericSrv* InService)
+bool ASimulationState::ServerCheckSetEntityStateRequest(const FROSSetEntityState_Request& InRequest)
```
why those checks are  necessary?

larryng-rapyuta replied:
To ensure that we don't perform the exact same request multiple times, this is more applicable to running this setup locally
as each robot's rosnode will receive the rosservice call and we will get 1 copy of the request for each robot

yuokamoto committed:
fix debug log (#76)

* fix debug log

Co-authored-by: Tadinu <ducanh.than@rapyuta-robotics.com>

Commit:
Simulation state add doxygen

Commit:
SimulationState.h doxygen cmt: service, action -> interface

Commit:
RROS2SimulationStateClient Rollback rpc call to Server's SimState methods
RRRobotBaseVehicle Rollback to fetching ClientTimeStamp + ClientRobotPose to pass to RPC ServerSetLinear/AngularVel()

Commit:
RRBaseActor bReplicates, bAlwaysRelevant true
UROS2Spawnable::OnComponentCreated() SetIsReplicated(true)

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Core/RRNetworkPlayerController.cpp:
```diff
@@ -0,0 +1,327 @@
+    PlayerName = InPlayerName;
+    if (SimStateClientROS2Node)
+    {
+        SimStateClientROS2Node->Namespace = InPlayerName;
```
what is this for?

larryng-rapyuta replied:
originally this was to set the server's name and namespace and let it replicate down, mainly to deal with timing issues from the editor

yuokamoto commented on Source/RapyutaSimulationPlugins/Private/Core/RRNetworkGameMode.cpp:
```diff
@@ -0,0 +1,49 @@
+    // pc's server name
+    networkPlayerController->SetName(pcName);
+    // pc's client local name
+    networkPlayerController->Rename(*pcName);
```
Do we need rename? SetName do not set both client and server?

larryng-rapyuta replied:
Actually the SetName and Rename functions each rename the player and the player controller's name in the editor, mainly so that we can see it easier in the editor and identify them

yuokamoto commented on Source/RapyutaSimulationPlugins/Public/Core/RRNetworkPlayerController.h:
```diff
@@ -0,0 +1,142 @@
+    // PLAYER
+    //! Controller's PlayerName, taken from [robotname] param in Engine build.
+    //! In Editor build, it will be automatically set from #PossessedPawn's name upon possession
+    UPROPERTY(EditAnywhere, BlueprintReadWrite, meta = (ExposeOnSpawn = "true"), Replicated)
```
Do we need PlayerName? Can't we use actor's name directly?

larryng-rapyuta replied:
there are some issues with the replication with the actor's name, it doesnt seem to replicate over correctly and generally tries to take the name when its spawn in the client (the default object name), its mainly why I ended up having to use PlayerName

yuokamoto replied:
I see.
btw, it seems post login happen before create simustateclient, which result in same name simstateclient ros2 node in each clients

[2022.08.03-09.49.57:777][886]LogRapyutaCore: Warning: [PC:1549518960] ROS2SimStateClient[NetworkPC1ROS2SimStateClient:239099872] created
[2022.08.03-09.49.57:777][886]LogRapyutaCore: Warning: ARRNetworkGameMode::PostLogin Logged-in PC[1] NetworkPC1

Commit:
RRNetworkPlayerController
ServerSetPlayerName() rem SimStateClientROS2Node->Namespace setting since PlayerName is meant for Robot's ROS2Node namespace
FindPawnToPossess() Use RobotUniqueName which is replicated for name matching

yuokamoto committed:
ARRNetworkPlayerController::FindPawnToPossess() Use APawn::GetPlayerState() (#77)

* add replicable target point

* update log

* use GetPlayerState to check pawn is controlled by playercontroller

yuokamoto committed:
Sync Game time in networkplayercontroller (#79)
